xtc/src/ledger.rs

```diff
@@ -5,54 +5,63 @@ use ic_cdk::export::candid::{CandidType, Principal};
 use ic_cdk::*;
 use ic_cdk_macros::*;
 use serde::*;
-use std::collections::hash_map::Entry;
 use std::collections::HashMap;
 
-pub struct Ledger(HashMap<Principal, u64>);
+#[derive(Default)]
+pub struct Ledger {
+
+    // stores the cycle balance hold by the principal
+    balances: HashMap<Principal, u64>,
 
-impl Default for Ledger {
-    fn default() -> Self {
-        Self(HashMap::new())
-    }
 }
 
 impl Ledger {
+
     pub fn archive(&mut self) -> Vec<(Principal, u64)> {
-        let map = std::mem::replace(&mut self.0, HashMap::new());
-        map.into_iter()
+        std::mem::take(&mut self.balances)
+            .into_iter()
             .filter(|(_, balance)| *balance > 0)
             .collect()
     }
 
     pub fn load(&mut self, archive: Vec<(Principal, u64)>) {
-        self.0 = archive.into_iter().collect();
-        self.0.reserve(25_000 - self.0.len());
+        self.balances = archive.into_iter().collect();
+        self.balances.reserve(25_000 - self.balances.len());
     }
 
     #[inline]
     pub fn balance(&self, account: &Principal) -> u64 {
-        match self.0.get(account) {
-            Some(balance) => *balance,
-            None => 0,
-        }
+        *(self.balances.get(account).unwrap_or(&0))
     }
 
     #[inline]
     pub fn deposit(&mut self, account: Principal, amount: u64) {
         StatsData::deposit(amount);
-        match self.0.entry(account) {
-            Entry::Occupied(mut e) => {
-                e.insert(*e.get() + amount);
-            }
-            Entry::Vacant(e) => {
-                e.insert(amount);
+        *(self.balances.entry(account).or_default()) += amount;
+    }
+
+    #[inline]
+    pub fn withdraw_erc20(&mut self, account: &Principal, amount: u64) -> Result<u64, ErrorDetails> {
+        let balance = match self.balances.get_mut(&account) {
+            Some(balance) if *balance >= amount => {
+                *balance -= amount;
+                *balance
             }
+            _ => return Err(InsufficientBalanceError.clone()),
+        };
+
+        if balance == 0 {
+            self.balances.remove(&account);
         }
+
+        StatsData::withdraw(amount);
+
+        Ok(balance)
     }
 
     #[inline]
     pub fn withdraw(&mut self, account: &Principal, amount: u64) -> Result<(), ()> {
```

Comment:
Result<(),()> is basically a boolean

Reply:
but `Result<(), ()>` allows usage of `?` syntax, which is nice :D